### DISCUSSION TITLE
Fulfillment deletion

### DISCUSSION THREAD

**Discussion by turacma** [medusajs/medusa#10020] [Feature Requests]
### Package.json file

```JSON
{
  "name": "ship",
  "version": "0.0.1",
  "description": "SHIP for Eight Sleep",
  "author": "Eight Sleep",
  "keywords": [
    "sqlite",
    "postgres",
    "typescript",
    "ecommerce",
    "headless",
    "medusa"
  ],
  "scripts": {
    "build": "medusa build",
    "start": "medusa start",
    "dev": "medusa develop",
    "test:integration:http": "TEST_TYPE=integration:http NODE_OPTIONS=--experimental-vm-modules jest --silent=false --runInBand --forceExit",
    "test:integration:modules": "TEST_TYPE=integration:modules NODE_OPTIONS=--experimental-vm-modules jest --silent --runInBand --forceExit",
    "test:unit": "TEST_TYPE=unit NODE_OPTIONS=--experimental-vm-modules jest --silent --runInBand --forceExit"
  },
  "dependencies": {
    "@medusajs/admin-sdk": "2.0.1",
    "@medusajs/api-key": "2.0.1",
    "@medusajs/auth": "2.0.1",
    "@medusajs/auth-emailpass": "2.0.1",
    "@medusajs/cache-inmemory": "2.0.1",
    "@medusajs/cart": "2.0.1",
    "@medusajs/cli": "2.0.1",
    "@medusajs/currency": "2.0.1",
    "@medusajs/customer": "2.0.1",
    "@medusajs/event-bus-local": "2.0.1",
    "@medusajs/event-bus-redis": "2.0.1",
    "@medusajs/framework": "2.0.1",
    "@medusajs/file": "2.0.1",
    "@medusajs/file-local": "2.0.1",
    "@medusajs/fulfillment": "2.0.1",
    "@medusajs/fulfillment-manual": "2.0.1",
    "@medusajs/inventory": "2.0.1",
    "@medusajs/js-sdk": "^2.0.1",
    "@medusajs/medusa": "2.0.1",
    "@medusajs/notification": "2.0.1",
    "@medusajs/notification-local": "2.0.1",
    "@medusajs/order": "2.0.1",
    "@medusajs/payment": "2.0.1",
    "@medusajs/pricing": "2.0.1",
    "@medusajs/product": "2.0.1",
    "@medusajs/promotion": "2.0.1",
    "@medusajs/region": "2.0.1",
    "@medusajs/sales-channel": "2.0.1",
    "@medusajs/stock-location": "2.0.1",
    "@medusajs/store": "2.0.1",
    "@medusajs/tax": "2.0.1",
    "@medusajs/types": "2.0.1",
    "@medusajs/user": "2.0.1",
    "@medusajs/workflow-engine-inmemory": "2.0.1",
    "express": "^4.21.1",
    "stripe": "^17.2.1"
  },
  "devDependencies": {
    "@mikro-orm/cli": "5.9.7",
    "@mikro-orm/core": "5.9.7",
    "@mikro-orm/migrations": "5.9.7",
    "@mikro-orm/postgresql": "5.9.7",
    "@stdlib/number-float64-base-normalize": "0.2.3",
    "@swc/core": "1.7.40",
    "@swc/jest": "^0.2.36",
    "@types/express": "^5.0.0",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.8.1",
    "@types/react": "^18.3.12",
    "jest": "^29.7.0",
    "@medusajs/test-utils": "2.0.1",
    "prop-types": "^15.8.1",
    "ts-node": "^10.9.2",
    "typescript": "^5.6.3"
  },
  "engines": {
    "node": ">=20"
  }
}
```

### Node.js version

v20.15.0

### Database and its version

Aurora Postgres

### Operating system name and version

Mac OS X 15.0.1 (24A348)

### What happended?

The fulfillment service offers no mechanism to remove fulfillments, only cancel them. 

While trying to figure out if the issue I reported in https://github.com/medusajs/medusa/issues/9964 was real I created a large number of fulfillments and no have no way to remove them. 

### Expected behavior

Fulfillment service would offer a `deleteFulfillments` method similar to other services.

### Actual behavior

Fulfillments can only be canceled

### Link to reproduction repo

See above

**Comment by sgirones** [comment]
hey @turacma, this behaviour is intentional today, but at the same time we are happy to revisit this product decision if needed.

Besides the test scenario you described, is there another scenario where being able to delete the fulfillments instead of only being able to cancel them, would be useful to you?

In any case, I'm moving this to a Discussion so we can continue this conversation.

**Reply by turacma** [reply]
@sgirones Thanks for the reply. I can't think of any other specific use case, it just seemed like a gap in providing customers full control of the data. Orders can be deleted, so why can't fulfillments?

**Reply by turacma** [reply]
@sgirones I may have discovered another use case, but perhaps you might have another suggestion on how to model. 

Some of the products we sell are fulfilled by a third party. We create orders with them and they will eventually ship. It is possible that those fulfillments may have to be sent to customers across multiple shipments, but we won't know that until well after we submit the request to them.

Given that there appears to be a 1:1 relationship between a shipment and fulfillment, my initial thought was when we discover this was delete and create 2 new fulfillments, but that wouldn't be possible. I'm guessing the suggested workflow would be to update the existing fulfillment to remove items then create a second one? Is it possible to have a long running workflow in the provider implementation that could ultimately create multiple fulfillments as shipments are created?

**Reply by turacma** [reply]
@sgirones bump on this topic as it's something I now have to solve as part of our pilot with Medusa.

**Reply by SalahAdDin** [reply]
> @sgirones bump on this topic as it's something I now have to solve as part of our pilot with Medusa.

Couldn't you make PR for solving it?

**Reply by turacma** [reply]
@SalahAdDin Sure, but I'm looking for feedback on an approach. I don't want to write a PR for something that medusa folks won't accept.
